File: python/problems/leetcode/0018_four_sum/solution.py
"""LeetCode 18: 4Sum."""

from __future__ import annotations


class Solution:
    def fourSum(self, nums: list[int], target: int) -> list[list[int]]:
        nums.sort()
        n = len(nums)
        res: list[list[int]] = []
        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i + 1, n):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                l, r = j + 1, n - 1
                while l < r:
                    s = nums[i] + nums[j] + nums[l] + nums[r]
                    if s < target:
                        l += 1
                    elif s > target:
                        r -= 1
                    else:
                        res.append([nums[i], nums[j], nums[l], nums[r]])
                        x, y = nums[l], nums[r]
                        while l < r and nums[l] == x:
                            l += 1
                        while l < r and nums[r] == y:
                            r -= 1
        return res


if __name__ == "__main__":
    out = Solution().fourSum([1, 0, -1, 0, -2, 2], 0)
    assert sorted(map(tuple, out)) == [
        (-2, -1, 1, 2), (-2, 0, 0, 2), (-1, 0, 0, 1),
    ]
    assert Solution().fourSum([2, 2, 2, 2, 2], 8) == [[2, 2, 2, 2]]
    print("0018 OK")

File: cpp/problems/leetcode/0018_four_sum/solution.cpp
// LeetCode 18
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = (int)nums.size();
        vector<vector<int>> res;
        for (int i = 0; i < n; ++i) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j < n; ++j) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                int l = j + 1, r = n - 1;
                while (l < r) {
                    long long s = (long long)nums[i] + nums[j] + nums[l] + nums[r];
                    if (s < target)
                        ++l;
                    else if (s > target)
                        --r;
                    else {
                        res.push_back({nums[i], nums[j], nums[l], nums[r]});
                        int xl = nums[l], yr = nums[r];
                        while (l < r && nums[l] == xl) ++l;
                        while (l < r && nums[r] == yr) --r;
                    }
                }
            }
        }
        return res;
    }
};

int main() {
    vector<int> a{1, 0, -1, 0, -2, 2};
    auto out = Solution().fourSum(a, 0);
    set<vector<int>> got(out.begin(), out.end());
    set<vector<int>> expect{
        {-2, -1, 1, 2}, {-2, 0, 0, 2}, {-1, 0, 0, 1},
    };
    assert(got == expect);
    vector<int> b{2, 2, 2, 2, 2};
    assert((Solution().fourSum(b, 8) == vector<vector<int>>{{2, 2, 2, 2}}));
    cout << "0018 OK" << endl;
    return 0;
}
